### PR TITLE
Various fixes with OpenACC backend for watch and derivimplicit

### LINKED ISSUE #379: Various issues while compiling mod files with OpenACC backend (ThreshDetect.mod orn.mod ampanmda.mod)
See below errors:

```
"x86_64/core/mod2c/ThreshDetect.cpp", line 258: warning: "inline" is not
          allowed
          static inline void net_receive_kernel_ThreshDetect(double t, Point_process* pnt, ThreshDetect_Instance* inst, NrnThread* nt, Memb_list* ml, int weight_index, double flag)  {
                 ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 258: error: a block-scope function
          may only have extern storage class
          static inline void net_receive_kernel_ThreshDetect(double t, Point_process* pnt, ThreshDetect_Instance* inst, NrnThread* nt, Memb_list* ml, int weight_index, double flag)  {
          ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 258: error: expected a ";"
          static inline void net_receive_kernel_ThreshDetect(double t, Point_process* pnt, ThreshDetect_Instance* inst, NrnThread* nt, Memb_list* ml, int weight_index, double flag)  {
                                                                                                                                                                                      ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 309: error: identifier "inst" is
          undefined
                  #pragma acc parallel loop present(inst, nrb) async(nt->stream_id)
                                                    ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 309: error: identifier "nrb" is
          undefined
                  #pragma acc parallel loop present(inst, nrb) async(nt->stream_id)
                                                          ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 319: warning: parsing restarts here
          after previous syntax error
                          Point_process* point_process = nt->pntprocs + offset;
                                                                              ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 320: error: identifier
          "point_process" is undefined
                          net_receive_kernel_ThreshDetect(t, point_process, inst, nt, ml, weight_index, flag);
                                                             ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 320: error: identifier "inst" is
          undefined
                          net_receive_kernel_ThreshDetect(t, point_process, inst, nt, ml, weight_index, flag);
                                                                            ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 320: error: identifier
          "weight_index" is undefined
                          net_receive_kernel_ThreshDetect(t, point_process, inst, nt, ml, weight_index, flag);
                                                                                          ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 320: error: identifier "flag" is
          undefined
                          net_receive_kernel_ThreshDetect(t, point_process, inst, nt, ml, weight_index, flag);
                                                                                                        ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 324: error: identifier
          "NetSendBuffer_t" is undefined
                  NetSendBuffer_t* nsb = ml->_net_send_buffer;
                  ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 324: error: identifier "ml" is
          undefined
                  NetSendBuffer_t* nsb = ml->_net_send_buffer;
                                         ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 325: error: expected a declaration
                  for (int i=0; i < nsb->_cnt; i++) {
                  ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 335: warning: parsing restarts here
          after previous syntax error
                  nsb->_cnt = 0;
                               ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 337: warning: this declaration has
          no storage class or type specifier
                  nrb->_displ_cnt = 0;
                  ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 337: error: expected a ";"
                  nrb->_displ_cnt = 0;
                     ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 338: warning: this declaration has
          no storage class or type specifier
                  nrb->_cnt = 0;
                  ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 338: error: variable "nrb" has
          already been defined
                  nrb->_cnt = 0;
                  ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 338: error: expected a ";"
                  nrb->_cnt = 0;
                     ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 339: error: expected a declaration
              }
              ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 345: error: identifier "nt" is
          undefined
              #pragma acc data present(nt, ml, ThreshDetect_global)
                                       ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 345: error: non variable found in
          parallel pragma list
              #pragma acc data present(nt, ml, ThreshDetect_global)
                                           ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 345: error: identifier
          "ThreshDetect_global" is undefined
              #pragma acc data present(nt, ml, ThreshDetect_global)
                                               ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 361: error: identifier "inst" is
          undefined
                      #pragma acc parallel loop present(inst, node_index, data, voltage, indexes, thread) async(nt->stream_id)
                                                        ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 361: error: identifier "node_index"
          is undefined
                      #pragma acc parallel loop present(inst, node_index, data, voltage, indexes, thread) async(nt->stream_id)
                                                              ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 361: error: identifier "data" is
          undefined
                      #pragma acc parallel loop present(inst, node_index, data, voltage, indexes, thread) async(nt->stream_id)
                                                                          ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 361: error: identifier "voltage" is
          undefined
                      #pragma acc parallel loop present(inst, node_index, data, voltage, indexes, thread) async(nt->stream_id)
                                                                                ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 361: error: identifier "indexes" is
          undefined
                      #pragma acc parallel loop present(inst, node_index, data, voltage, indexes, thread) async(nt->stream_id)
                                                                                         ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 361: error: identifier "thread" is
          undefined
                      #pragma acc parallel loop present(inst, node_index, data, voltage, indexes, thread) async(nt->stream_id)
                                                                                                  ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 361: error: identifier "nt" is
          undefined
                      #pragma acc parallel loop present(inst, node_index, data, voltage, indexes, thread) async(nt->stream_id)
                                                                                                                ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 365: warning: parsing restarts here
          after previous syntax error
                          double v = voltage[node_id];
                                                     ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 366: warning: this declaration has
          no storage class or type specifier
                          net_send_buffering(ml->_net_send_buffer, 0, inst->tqitem[2*pnodecount+id], -1, inst->point_process[1*pnodecount+id], nt->_t+0.0, 1.0);
                          ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 366: error: expected a ")"
                          net_send_buffering(ml->_net_send_buffer, 0, inst->tqitem[2*pnodecount+id], -1, inst->point_process[1*pnodecount+id], nt->_t+0.0, 1.0);
                                                                 ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 367: error: expected a declaration
                      }
                      ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 375: error: non variable found in
          parallel pragma list
              #pragma acc data present(nt, ml, ThreshDetect_global)
                                       ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 375: error: non variable found in
          parallel pragma list
              #pragma acc data present(nt, ml, ThreshDetect_global)
                                           ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 375: error: identifier
          "ThreshDetect_global" is undefined
              #pragma acc data present(nt, ml, ThreshDetect_global)
                                               ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 388: error: identifier "inst" is
          undefined
                  #pragma acc parallel loop present(inst, node_index, data, voltage, indexes, thread) async(nt->stream_id)
                                                    ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 388: error: identifier "node_index"
          is undefined
                  #pragma acc parallel loop present(inst, node_index, data, voltage, indexes, thread) async(nt->stream_id)
                                                          ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 388: error: identifier "data" is
          undefined
                  #pragma acc parallel loop present(inst, node_index, data, voltage, indexes, thread) async(nt->stream_id)
                                                                      ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 388: error: identifier "voltage" is
          undefined
                  #pragma acc parallel loop present(inst, node_index, data, voltage, indexes, thread) async(nt->stream_id)
                                                                            ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 388: error: identifier "indexes" is
          undefined
                  #pragma acc parallel loop present(inst, node_index, data, voltage, indexes, thread) async(nt->stream_id)
                                                                                     ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 388: error: identifier "thread" is
          undefined
                  #pragma acc parallel loop present(inst, node_index, data, voltage, indexes, thread) async(nt->stream_id)
                                                                                              ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 419: warning: parsing restarts here
          after previous syntax error
              hoc_register_net_send_buffering(mech_type);
                                                        ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 420: warning: this declaration has
          no storage class or type specifier
              hoc_register_var(hoc_scalar_double, hoc_vector_double, NULL);
              ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 420: error: identifier
          "hoc_scalar_double" is undefined
              hoc_register_var(hoc_scalar_double, hoc_vector_double, NULL);
                               ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 420: error: expected a ")"
              hoc_register_var(hoc_scalar_double, hoc_vector_double, NULL);
                                                ^

"x86_64/core/mod2c/ThreshDetect.cpp", line 421: error: expected a declaration
          }
          ^

39 errors detected in the compilation of "x86_64/core/mod2c/ThreshDetect.cpp".
make: *** [x86_64/core/build/ThreshDetect.o] Error 2
make: *** Waiting for unfinished jobs....
"x86_64/core/mod2c/ampanmda.cpp", line 608: error: identifier "IonCurVar" is
          undefined
          IonCurVar ionvar;
          ^

"x86_64/core/mod2c/ampanmda.cpp", line 609: error: identifier
          "check_mgblock_AmpaNmda" is undefined
          check_mgblock_AmpaNmda(id, pnodecount, inst, data, indexes, thread, nt, v);
          ^

2 errors detected in the compilation of "x86_64/core/mod2c/ampanmda.cpp".
make: *** [x86_64/core/build/ampanmda.o] Error 2
```

```
 -> Compiling x86_64/core/mod2c/fi_stdp.cpp
"x86_64/core/mod2c/ampanmda.cpp", line 608: error: identifier "IonCurVar" is
          undefined
          IonCurVar ionvar;
          ^

"x86_64/core/mod2c/ampanmda.cpp", line 609: error: identifier
          "check_mgblock_AmpaNmda" is undefined
          check_mgblock_AmpaNmda(id, pnodecount, inst, data, indexes, thread, nt, v);
          ^
```

```
PGCC-S-0155-External variables used in acc routine need to be in #pragma acc create() - _ZN10coreneuron2dtE (x86_64/core/mod2c/orn.cpp: 512)
PGCC-S-0155-Accelerator region ignored; see -Minfo messages  (x86_64/core/mod2c/orn.cpp)
PGCC/x86-64 Linux 19.10-0: compilation completed with severe errors
make: *** [x86_64/core/build/orn.o] Error 2
```

### PR DESCRIPTION
**Pull request by pramodk**
* Various fixes with OpenACC backend for watch & derivimplicit
     - artificial cells shouldn't use cuda memory allocation routines (but on cpu)
     - artificial cells shouldn't be compiled for GPU
     - use backend specific abort routine as abort is not supported on gpus
     - transfer routine to copy newtonspace on gpu
 * Azure CI changes
  - use python 3.7
  - use delocate on OS X instead of auditwheel
* Add test for watch statement

    Fixes #379 #115 #375

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -125,16 +125,23 @@ jobs:
...
       python3 -m venv venv
       source ./venv/bin/activate
       pip install --upgrade pip
-      pip install dist/*.whl
+      pip install wheelhouse/*.whl
       # TODO : change this when we fix the issue with nmodl module
       export PYTHONPATH=$(pwd)/venv/lib/python3.8/site-packages:$PYTHONPATH
-      nmodl $(Build.Repository.LocalPath)/test/integration/mod/cabpump.mod sympy --analytic
+      cd $(Build.Repository.LocalPath)/test/integration/mod
+      nmodl cabpump.mod sympy --analytic
```
[the remark this answers is not in the document]
Comment (pramodk) [reply azure-pipelines.yml:145]: oh yes, I had `nrnivmodl-core` in my mind.

File: src/codegen/codegen_c_visitor.cpp [hunk 2]
```diff
@@ -3222,9 +3250,15 @@ void CodegenCVisitor::print_nrn_init(bool skip_init_check) {
         int list_num = info.derivimplicit_list_num;
         // clang-format off
         printer->add_line("*deriv{}_advance(thread) = 0;"_format(list_num));
-        printer->add_line("if (*newtonspace{}(thread) == NULL) {}"_format(list_num, "{"));
-        printer->add_line("    *newtonspace{}(thread) = nrn_cons_newtonspace({}, pnodecount);"_format(list_num, nequation));
-        printer->add_line("    thread[dith{}()].pval = makevector(2*{}*pnodecount*sizeof(double));"_format(list_num, nequation));
+        printer->add_line("auto ns = newtonspace{}(thread);"_format(list_num));
+        printer->add_line("auto& th = thread[dith{}()];"_format(list_num));
+
+        printer->add_line("if (*ns == NULL) {");
```
[the remark this answers is not in the document]
Comment (pramodk) [reply src/codegen/codegen_c_visitor.cpp:3256]: yeah minor.. but I will change NULL to nullptr.

File: src/codegen/codegen_c_visitor.cpp [hunk 3]
```diff
@@ -3281,14 +3315,16 @@ void CodegenCVisitor::print_watch_activate() {
     auto inst = "{}* inst"_format(instance_struct());
 
     printer->start_block(
-        "static void nrn_watch_activate({}, int id, int pnodecount, int watch_id) "_format(inst));
+        "static void nrn_watch_activate({}, int id, int pnodecount, int watch_id, double v, bool &watch_remove) "_format(
+            inst));
 
     // initialize all variables only during first watch statement
-    printer->add_line("if (watch_id == 0) {");
+    printer->add_line("if (watch_remove == false) {");
```
[the remark this answers is not in the document]
Comment (pramodk) [reply src/codegen/codegen_c_visitor.cpp:3322]: we do run that in parallel but `watch_remove` is an argument to function and that is local variable in each parallel invocation. So there won't be any race condition.

**Review by ohm314 — APPROVED** [review]
while I don't really know much about the WATCH statement, the code looks otherwise good to me.

**Review by cattabiani — APPROVED** [review]

**Pull request merged**